[Macros] Update plugin search options serialization

rintaro opened the pull request:
Previously plugin search options were serialized for each option kind. Instead serialize them in the order specified.

rdar://110903149

rintaro committed:
[Macros] Update plugin search options serialization

Previously plugin search options were serialized for each option kind.
Instead serialize them in the order specified.

rintaro commented:
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

rintaro commented on lib/Serialization/ModuleFileSharedCore.cpp:
```diff
@@ -126,18 +126,27 @@ static bool readOptionsBlock(llvm::BitstreamCursor &cursor,
+      options_block::ResilienceStrategyLayout::readRecord(scratch, optKind);
+      switch (PluginSearchOptionKind(optKind)) {
+      case PluginSearchOptionKind::PluginPath:
+        optStr = "-plugin-path";
```
Not sure about this. But I didn't wanted to introduce another "kind" enum just for `swift/Serialization/Validation.h`

DougGregor replied:
The fact that LLDB needs to match on these strings feels a little brittle. Would we be better off exporting `PluginSearchOptionKind` into a header and having that in the list of plugin search options?

rintaro replied:
👍 Added a commit.
Made `PluginSearchOption` a struct with `PluginSearchOption::Kind` in `SearchPathOptions.h`

xymus commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -651,6 +651,16 @@ enum class MacroIntroducedDeclNameKind : uint8_t {
+  LoadPluginLibrary,
+  LoadPluginExecutable,
+};
+using PluginSearchOptionKindField = BCFixed<3>;
```
It looks like a `BCFixed<2>` should be enough for the current enum size.

xymus reviewed (approved):
The change looks good!

rintaro committed:
[Macros] Make 'PluginSearchOption' a external union

Create a 'Kind' enum so that deserialization can use the kind instead of
a string option name.

rintaro commented:
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7003
@swift-ci Please smoke test

bnbarham reviewed (approved):
LGTM other than @xymus comment

rintaro commented:
`BCFixed<3>` is for possible future addition of another options without incrementing module format version.
Since this record is _not_ going to be used hundreds of times for each module, @xymus and I agreed we keep it 3 bits.
